aiohttp/connector.py

```diff
@@ -982,25 +980,38 @@ def _get_ssl_context(self, req: ClientRequest) -> Optional[SSLContext]:
             3. if verify_ssl is False in req, generate a SSL context that
                won't verify
         """
-        if req.is_ssl():
-            if ssl is None:  # pragma: no cover
-                raise RuntimeError("SSL is not supported.")
-            sslcontext = req.ssl
-            if isinstance(sslcontext, ssl.SSLContext):
-                return sslcontext
-            if sslcontext is not True:
-                # not verified or fingerprinted
-                return self._make_ssl_context(False)
-            sslcontext = self._ssl
-            if isinstance(sslcontext, ssl.SSLContext):
-                return sslcontext
-            if sslcontext is not True:
-                # not verified or fingerprinted
-                return self._make_ssl_context(False)
-            return self._make_ssl_context(True)
-        else:
+        if not req.is_ssl():
             return None
 
+        if ssl is None:  # pragma: no cover
+            raise RuntimeError("SSL is not supported.")
+        sslcontext = req.ssl
+        if isinstance(sslcontext, ssl.SSLContext):
+            return sslcontext
+        if sslcontext is not True:
+            # not verified or fingerprinted
+            return await self._make_or_get_cached_ssl_context(False)
+        sslcontext = self._ssl
+        if isinstance(sslcontext, ssl.SSLContext):
+            return sslcontext
+        if sslcontext is not True:
+            # not verified or fingerprinted
+            return await self._make_or_get_cached_ssl_context(False)
+        return await self._make_or_get_cached_ssl_context(True)
+
+    async def _make_or_get_cached_ssl_context(self, verified: bool) -> SSLContext:
+        """Create or get cached SSL context."""
+        if verified in self._made_ssl_context:
+            return self._make_ssl_context(verified)
+        # _make_ssl_context does blocking I/O to load certificates
+        # from disk, so we run it in a separate thread.
+        async with self._make_ssl_context_lock.setdefault(verified, asyncio.Lock()):
+            context = await self._loop.run_in_executor(
+                None, self._make_ssl_context, verified
+            )
+            self._made_ssl_context.add(verified)
+        return context
```

Comment:
Would it be a bit cleaner to do something along the lines of:

```suggestion
        try:
            future = self._make_ssl_context[verified]
        except KeyError:
            future = asyncio.Future()
            self._make_ssl_context[verified] = future
            try:
                result = await self._loop.run_in_executor(
                    None, self._make_ssl_context
                )
            except Exception as e:
                future.set_exception(e)
                self._make_ssl_context.pop(verified)
            else:
                future.set_result(result)
        return await future
```

The other function no longer needs to have a cache.

Reply:
I think the original code could result in many executor calls just to run a cached function. It also just feels a little odd having a dict to check if the function is already cached, when you could instead store the result in there.

Reply:
I think we would need to worry about cancellation with this approach.... maybe we need to check for BaseException.  Need to think on this